### DISCUSSION TITLE
How to get more granular control for window rules?

### DISCUSSION THREAD

**Discussion by szfkamil** [asmvik/yabai#2347] [Q&A]
I have been trying to create a separate rule for the Quick Add window for Todoist and the main app window.

> The QuickAdd window moves to the Todoist window's space, even if it is in a different space and loses focus. It gets 'glued' to the Todoist window. In other words, it strives to float on top of the Todoist window. When repositioned, it always changes its position back. 

This issue to the grid and space rule. I queried properties of both windows used `diff` to look for stable properties that differentiate the both windows, and have found that the Quick Add is always an exact match of “Todoist”, whereas the latter is always “Todoist - (something else)”

So, I created those two rules:

```
yabai -m rule --add app="^Todoist$" title="^Todoist$" label="Quick Add" manage=off sticky=on
# Cannot add grid rule due to limitation. Grid rule gets applied for the Quick
# Add window.
yabai -m rule --add app="Todoist" title="Todoist" label="Todoist window" manage=off space=sm 

```

It seems that yabai does not support:

- applying rule based on frame dimensions (Quick Add is always the same size)
- complex regex patterns like negative lookahead directly

```
invalid regex pattern '^(?!Todoist$).*' for key 'title'
invalid regex pattern '^((?!Todoist$).)*$' for key 'title'
```

So based on where I am at, I cannot differentiate between those two windows of the app. 

What can I do in order to more granularly differentiate between app windows and set custom rules for each? This is the general problem I have, and I would appreciate some help on achieving my specific end with the Todoist windows.

Thanks!

**Comment by Bellavene** [comment]
You need to query the window role and subrole.
After just make a rule in .yabairc with corresponding roles.

```
yabai -m rule --add app="^Adobe.*$" \
         role="AXLayoutArea" \
         subrole="AXFloatingWindow" \
         manage=on \
         display=^1 \
         grid=1:1:0:0:1:1
```
